sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala

```diff
@@ -800,4 +800,39 @@ class SparkSqlParserSuite extends AnalysisTest with SharedSparkSession {
         start = 0,
         stop = 63))
   }
+
+  def getMemoryUsage: Long = {
+    System.gc()
+    Runtime.getRuntime.totalMemory() - Runtime.getRuntime.freeMemory()
+  }
+
+  def awfulQuery(depth: Int): String = {
+    if (depth == 0) {
+      s"rand()"
+    } else {
+      s"case when ${awfulQuery(depth - 1)} > 0.5 " +
+      s"then ${awfulQuery(depth - 1)} " +
+      s"else ${awfulQuery(depth - 1)} " +
+      "end"
+    }
+  }
+
+  test("release ANTLR cache after parsing") {
+    val baselineMemory = getMemoryUsage // On my system, this is about 61 MiB
+    parser.parsePlan(s"select ${awfulQuery(8)} from range(10)")
+    val estimatedCacheOverhead = getMemoryUsage - baselineMemory // about  119 MiB
```

Comment:
I'm worrying about the possibility where this test case might introduce a new flakiness in our GitHub Action environment. Especially, Apple Silicon environment. Do you think this test case clears up the leftovers after completing this test suite?

Reply:
I worry a little about flakiness too, but the thresholding is intended to prevent that from being an issue.

Why especially Apple Silicon?

What leftovers are you referring to?  The only thing I can think of is the ANTLR cache, and this test is explicitly verifying that the ANTLR cache is cleared.  So I think yes, all leftovers are cleared?